bolt/lib/Core/DIEBuilder.cpp

```diff
@@ -378,11 +377,13 @@ getUnitForOffset(DIEBuilder &Builder, DWARFContext &DWCtx,
   return nullptr;
 }
 
-uint32_t DIEBuilder::computeDIEOffset(const DWARFUnit &CU, DIE &Die,
+uint32_t DIEBuilder::computeDIEOffset(DWARFUnit &CU, DIE &Die,
```

Comment:
If the method computes DIE offset, it's better to keep the CU parameter a constant ref, and do the modifications outside of this method.

Reply:
I removed const because DWARF5AcceleratorTable::addUnit invokes StrOffsetsWriter.initialize(Unit); which invokes DebugStrOffsetsWriter::initialize, which invokes Unit.getStringOffsetsTableContribution() that is non const. Because it invokes extractDIEsIfNeeded.

Reply:
I'm strongly in favor of keeping CU a const ref since semantically `computeDIEOffset` should not change the input CU (or DIE). Please make modifications to the CU outside of this method.

Reply:
I am not making changes to CU. I am invoking none const APIs from it.

Reply:
Can you then add a const API so you don't have to drop const qualifier? As a separate diff.

Reply:
As I mentioned the issue is getStringOffsetsTableContribution which invoked extractDIEsIfNeeded. Which parses dwarf and sets StringOffsetsTableContribution, which that API returns. So it can't be const.

Reply:
Issue is 

```
 const std::optional<StrOffsetsContributionDescriptor> &
  getStringOffsetsTableContribution() {
    extractDIEsIfNeeded(true /*CUDIeOnly*/);
    return StringOffsetsTableContribution;
  }
```
It parses debug information and sets StringOffsetsTableContribution first time it's called.

I can add 

```
  const std::optional<StrOffsetsContributionDescriptor> &
  getStringOffsetsTableContribution() const {
    assert(StringOffsetsTableContribution);
    return StringOffsetsTableContribution;
  } 
```

But then implementation will rely on extractDIEsIfNeeded being invoked for this cu at some point. Probably will, but this seems flaky to me.

Reply:
I tried with change on an internal service of reasonable size. No issues, but having basically the same API with quite different behavior based on const seems like just asking for trouble.

Reply:
I hope you understand why I'm raising this issue – that if the method is called `computeDIEOffset`, it's reasonable that CU is passed as a const ref.

Dropping the const qualifier is a symptom of a larger issue: that a parameter that's not supposed to be modified starts to be modified. I understand that it's not possible/practical to add a const-qualified method, that's ok. I suggested moving the modifications to CU/calling non-const stuff outside of this method – what do you think about that?
As a final resort, you can rename the method to `computeDIEOffsetAndAddAccelTableEntry` and drop const.

Reply:
I see your point. The reason I put adding entries here to avoid traversing all these CUs/DIEs again and wasting resources. 
I can get StringOffsetsTableContribution before the call, but I don't think it fundamentally addresses your concern. Why are we passing StrOffsetsContributionDescriptor into computeDIEOffset?

So I think the option you mentioned of renaming to something else would be the way to go. Maybe something more generic processAndFinalizeDIEs? In case down the road we add more stuff here. For example once ability to delete dies is added I think computing GDB Index will need to move out of main updatedebuginfo.